mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp

```diff
@@ -52,6 +52,82 @@ template <typename Ty> static Ty getMemRefElementType(Type ty) {
   return dyn_cast<Ty>(mt.getElementType());
 }
 
+/// Returns true if \p op may have an effect of type \tparam EffectTys on \p
+/// value.
+template <typename... EffectTys>
+static bool mayHaveEffect(Operation &op, Value value) {
+  if (auto memOp = dyn_cast<MemoryEffectOpInterface>(op)) {
+    if (hasEffect<EffectTys...>(&op, value))
+      return true;
+    // If the op does not have recursive side effects, then it does not have
+    // such effects in the value.
+    if (!op.hasTrait<OpTrait::HasRecursiveMemoryEffects>())
+      return false;
+  } else if (!op.hasTrait<OpTrait::HasRecursiveMemoryEffects>()) {
+    // Otherwise, if the op does not implement the memory effect interface and
+    // it does not have recursive side effects, then we safely assume it has the
+    // queried effects on the value.
+    return true;
+  }
+
+  return llvm::any_of(op.getRegions(), [&](Region &region) {
+    return llvm::any_of(region.getOps(), [&](Operation &op) {
+      return mayHaveEffect<EffectTys...>(op, value);
+    });
+  });
+}
+
+/// Test \p id is default constructed and it is not modified.
+static bool isDefaultConstructedID(Value id) {
+  // Check it is default constructed
+  auto constructor = id.getDefiningOp<sycl::SYCLIDConstructorOp>();
+  if (!constructor || !constructor.getArgs().empty())
+    return false;
+
+  // Check it has not been modified after construction
+  return llvm::none_of(id.getUsers(), [=](Operation *op) {
```

Comment:
There could be aliased memrefs that may write to the allocated id after it has been created. To do this safely I think the transformation would have to prove that the id is not aliased to any other value, or (better) that any potentially alias is not written to.

Reply:
True, as we don't have analysis information available here, I implemented a more restrictive check using `sycl` domain ops.